File: src/Manipulators/Orientation.php
<?php

namespace AndriesLouw\imagesweserv\Manipulators;

use AndriesLouw\imagesweserv\Manipulators\Helpers\Utils;
use Jcupitt\Vips\Image;

/**
 * @property int $rotation
 * @property bool $flip
 * @property bool $flop
 */
class Orientation extends BaseManipulator
{
    /**
     * Perform orientation image manipulation.
     *
     * @param  Image $image The source image.
     *
     * @return Image The manipulated image.
     */
    public function run(Image $image): Image
    {
        // Rotate if required.
        if ($this->rotation !== 0) {
            // Need to copy to memory, we have to stay seq.
            $image = $image->copyMemory()->rot('d' . $this->rotation);
        }

        // Flip (mirror about Y axis) if required.
        if ($this->flip) {
            $image = $image->flipver();
        }

        // Flop (mirror about X axis) if required.
        if ($this->flop) {
            $image = $image->fliphor();
        }

        // Remove EXIF Orientation from image, if any
        if ($image->typeof(Utils::VIPS_META_ORIENTATION) !== 0) {
            $image->remove(Utils::VIPS_META_ORIENTATION);
        }

        return $image;
    }
}

File: src/weserv/manipulators/orientation.lua
--- Orientation manipulator
-- @module orientation
local manipulator = {}

--- Should this manipulator process the image?
-- @return Boolean indicating if we should process the image.
function manipulator.should_process(_)
    -- Always process the image via this manipulator.
    return true
end

--- Perform orientation image manipulation.
-- @param image The source image.
-- @param args The URL query arguments.
-- @return The manipulated image.
function manipulator.process(image, args)
    -- Rotate if required.
    if args.rotation ~= 0 then
        -- Need to copy to memory, we have to stay seq.
        image = image:copy_memory():rot("d" .. args.rotation)
    end

    -- Flip (mirror about Y axis) if required.
    if args.flip then
        image = image:flip("vertical")
    end

    -- Flop (mirror about X axis) if required.
    if args.flop then
        image = image:flip("horizontal")
    end

    -- Remove EXIF Orientation from image, if any
    if image:get_typeof("orientation") ~= 0 then
        image:remove("orientation")
    end

    return image
end

return manipulator